src/main/groovy/com/deploygate/gradle/plugins/DeployGatePlugin.groovy

```diff
@@ -36,71 +46,133 @@ class DeployGatePlugin implements Plugin<Project> {
+            task.description = "Execute all custom upload-aab tasks"
 
-        GradleCompat.configureEach(project.deploygate.deployments) { NamedDeployment deployment ->
-            processor.addVariantOrCustomName(deployment.name)
+            task.group = DeployGateTaskFactory.GROUP_NAME
         }
-    }
 
-    private void onProjectEvaluated(Project project) {
-        processor.declaredNames.forEach { variantOrCustomName ->
-            processor.registerDeclarationAwareUploadApkTask(variantOrCustomName)
-            processor.registerDeclarationAwareUploadAabTask(variantOrCustomName)
-        }
+        project.deploygate.deployments.configureEach { NamedDeployment d ->
+            project.tasks.named(DeployGateTaskFactory.SUFFIX_APK_TASK_NAME).configure { task ->
+                task.dependsOn(DeployGateTaskFactory.uploadApkTaskName(d.name))
+            }
+
+            project.tasks.named(DeployGateTaskFactory.SUFFIX_AAB_TASK_NAME).configure { task ->
+                task.dependsOn(DeployGateTaskFactory.uploadAabTaskName(d.name))
+            }
+
+            project.tasks.register(DeployGateTaskFactory.uploadApkTaskName(d.name), UploadApkTask) { task ->
+                final NamedDeployment deployment = project.deploygate.findDeploymentByName(d.name)
+
+                if (!deployment.skipAssemble) {
+                    task.logger.debug("${d.name} required assmble but ignored")
+                }
+
+                task.variantName = d.name
+                task.dependsOn(Constants.LOGIN_TASK_NAME)
+
+                task.configuration = UploadApkTask.createConfiguration(deployment, new DefaultPresetApkInfo(d.name))
+                task.applyTaskProfile()
+            }
+
+            project.tasks.register(DeployGateTaskFactory.uploadAabTaskName(d.name), UploadAabTask) { task ->
+                final NamedDeployment deployment = project.deploygate.findDeploymentByName(d.name)
+
+                if (!deployment.skipAssemble) {
+                    task.logger.debug("${d.name} required assmble but ignored")
+                }
 
-        processor.registerAggregatedDeclarationAwareUploadApkTask(processor.declaredNames)
-        processor.registerAggregatedDeclarationAwareUploadAabTask(processor.declaredNames)
+                task.variantName = d.name
+                task.dependsOn(Constants.LOGIN_TASK_NAME)
 
-        if (!processor.canProcessVariantAware()) {
-            project.logger.warn("DeployGate Gradle Plugin is stopped because Android Gradle Plugin must be applied before.")
-            return
+                task.configuration = UploadAabTask.createConfiguration(deployment, new DefaultPresetAabInfo(d.name))
+                task.applyTaskProfile()
+            }
         }
 
-        GradleCompat.configureEach(project.android.applicationVariants) { /* ApplicationVariant */ variant ->
-            def variantProxy = new IApplicationVariantImpl(variant)
+        AndroidGradlePlugin.ifPresent(project) {
```

Comment:
[question]
why we need to check `ifPresent()` here?
Call `init()` method after `GradleCompat.init()` immediately in existing codes.

Reply:
@satsukies 
It's a refactoring result of transitive references to the existence of Android Gradle Plugin. ref: https://github.com/DeployGate/gradle-deploygate-plugin/pull/162/files

All plugins are resolved in Project#afterEvaluate but not in the-apply time so Project#hasPlugin returns false if DeployGate plugin is applied before Android Gradle Plugin. It means we can initialize Android Gradle Plugin related code immediately if AGP already exist, otherwise we have to delay the initialization till AGP is applied. `ifPresent` does this.

Reply:
Thank you for the explanations. I understand 👍